File: kid-ios/MyWorld/Parent/AlbumView.swift
import SwiftUI

/// Memorabilia album, organized into folders to match how the family thinks
/// about their tiles rather than a flat scroll:
///
///   Album → Section folder → Tile → all versions of that tile, newest first
///
/// Sections collapse into four parent-friendly buckets:
///   People       (people)
///   Words        (nouns + needs — the "things, adjectives, phrases" bucket)
///   Verbs        (verbs)
///   Celebrations (the holiday/birthday Events track)
struct AlbumView: View {
    @Environment(AuthManager.self) private var auth

    @State private var tiles: [APIClient.AlbumTile] = []
    @State private var loaded = false
    @State private var errorText: String?

    private let api = APIClient()

    var body: some View {
        Group {
            if let e = errorText {
                Text(e).font(.footnote).foregroundStyle(.red).padding()
            } else if !loaded {
                ProgressView("Loading…").frame(maxWidth: .infinity, alignment: .center).padding(.top, 60)
            } else if tiles.isEmpty {
                Text("No pictures yet. As tiles get new art, the old versions are kept here forever.")
                    .font(.footnote).foregroundStyle(.secondary).padding(24)
            } else {
                folderList
            }
        }
        .background(Color(hex: "#fff7fb"))
        .navigationTitle("Album")
        .navigationBarTitleDisplayMode(.inline)
        .task {
            do {
                tiles = try await api.albumByTile(childId: auth.childSlug, limit: 600)
            } catch {
                errorText = "Could not load the album: \(error.localizedDescription)"
            }
            loaded = true
        }
    }

    private var folderList: some View {
        ScrollView {
            VStack(spacing: 14) {
                ForEach(AlbumFolder.allCases) { folder in
                    let tilesInFolder = tiles.filter { folder.matches(section: $0.section) }
                    if !tilesInFolder.isEmpty {
                        NavigationLink {
                            AlbumFolderView(folder: folder, tiles: tilesInFolder)
                        } label: {
                            folderRow(folder: folder, tiles: tilesInFolder)
                        }
                        .buttonStyle(.plain)
                    }
                }
            }
            .padding(16)
        }
    }

    private func folderRow(folder: AlbumFolder, tiles: [APIClient.AlbumTile]) -> some View {
        let pictureCount = tiles.reduce(0) { $0 + (1 + $1.history.count) }
        // Cover stack: up to 3 thumbnails of the most recent tiles, fanned out.
        let covers = Array(tiles.prefix(3)).compactMap { $0.current?.blobKey ?? $0.history.first?.blobKey }
        return HStack(spacing: 14) {
            ZStack {
                ForEach(Array(covers.enumerated()), id: \.offset) { i, key in
                    MediaImage(blobKey: key)
                        .frame(width: 60, height: 60)
                        .clipShape(RoundedRectangle(cornerRadius: 10))
                        .overlay(RoundedRectangle(cornerRadius: 10).stroke(.white, lineWidth: 2))
                        .rotationEffect(.degrees(Double(i - 1) * 6))
                        .offset(x: CGFloat(i - 1) * 4)
                }
                if covers.isEmpty {
                    RoundedRectangle(cornerRadius: 10)
                        .fill(Color(hex: folder.tint).opacity(0.18))
                        .frame(width: 60, height: 60)
                        .overlay(Image(systemName: folder.icon).foregroundStyle(Color(hex: folder.tint)))
                }
            }
            .frame(width: 78)

            VStack(alignment: .leading, spacing: 3) {
                Text(folder.title)
                    .font(.system(size: 18, weight: .bold, design: .rounded))
                    .foregroundStyle(Color(hex: "#1f2937"))
                Text("\(tiles.count) tile\(tiles.count == 1 ? "" : "s") · \(pictureCount) picture\(pictureCount == 1 ? "" : "s")")
                    .font(.system(size: 12))
                    .foregroundStyle(Color(hex: "#6b7280"))
            }
            Spacer()
            Image(systemName: "chevron.right")
                .font(.system(size: 14, weight: .semibold))
                .foregroundStyle(Color(hex: "#c9b3bf"))
        }
        .padding(14)
        .background(.white, in: RoundedRectangle(cornerRadius: 18))
        .overlay(RoundedRectangle(cornerRadius: 18).stroke(Color(hex: "#f3c6da"), lineWidth: 1))
    }
}

/// Album → folder. Sections collapse into four parent-friendly buckets so the
/// top level fits on one screen.
enum AlbumFolder: String, CaseIterable, Identifiable {
    case people, words, verbs, celebrations
    var id: String { rawValue }
    var title: String {
        switch self {
        case .people:       return "People"
        case .words:        return "Words"
        case .verbs:        return "Verbs"
        case .celebrations: return "Celebrations"
        }
    }
    var subtitle: String {
        switch self {
        case .people:       return "Family, friends, helpers"
        case .words:        return "Things, adjectives, phrases"
        case .verbs:        return "Actions and feelings"
        case .celebrations: return "Birthdays and holidays"
        }
    }
    var icon: String {
        switch self {
        case .people:       return "person.2.fill"
        case .words:        return "tag.fill"
        case .verbs:        return "figure.run"
        case .celebrations: return "sparkles"
        }
    }
    var tint: String {
        switch self {
        case .people:       return "#ec4899"
        case .words:        return "#0ea5e9"
        case .verbs:        return "#16a34a"
        case .celebrations: return "#f59e0b"
        }
    }
    func matches(section: String?) -> Bool {
        let s = (section ?? "").lowercased()
        switch self {
        case .people:       return s == "people"
        case .words:        return s == "nouns" || s == "needs" || s.isEmpty   // unfiled lands here too
        case .verbs:        return s == "verbs"
        case .celebrations: return s == "events"
        }
    }
}

/// One folder open: tiles in that section listed by name with a thumbnail
/// stack showing current + how many older versions there are.
private struct AlbumFolderView: View {
    let folder: AlbumFolder
    let tiles: [APIClient.AlbumTile]

    var body: some View {
        ScrollView {
            VStack(spacing: 10) {
                ForEach(tiles) { tile in
                    NavigationLink {
                        AlbumTileView(tile: tile)
                    } label: {
                        tileRow(tile)
                    }
                    .buttonStyle(.plain)
                }
            }
            .padding(16)
        }
        .background(Color(hex: "#fff7fb"))
        .navigationTitle(folder.title)
        .navigationBarTitleDisplayMode(.large)
    }

    private func tileRow(_ tile: APIClient.AlbumTile) -> some View {
        HStack(spacing: 12) {
            if let key = tile.current?.blobKey ?? tile.history.first?.blobKey {
                MediaImage(blobKey: key)
                    .frame(width: 56, height: 56)
                    .clipShape(RoundedRectangle(cornerRadius: 12))
            } else {
                RoundedRectangle(cornerRadius: 12)
                    .fill(Color(hex: "#fce4ec"))
                    .frame(width: 56, height: 56)
            }
            VStack(alignment: .leading, spacing: 2) {
                Text(tile.label ?? "Untitled")
                    .font(.system(size: 16, weight: .semibold, design: .rounded))
                    .foregroundStyle(Color(hex: "#1f2937"))
                let total = 1 + tile.history.count
                Text("\(total) picture\(total == 1 ? "" : "s")")
                    .font(.system(size: 12))
                    .foregroundStyle(Color(hex: "#6b7280"))
            }
            Spacer()
            Image(systemName: "chevron.right")
                .font(.system(size: 13, weight: .semibold))
                .foregroundStyle(Color(hex: "#c9b3bf"))
        }
        .padding(12)
        .background(.white, in: RoundedRectangle(cornerRadius: 14))
        .overlay(RoundedRectangle(cornerRadius: 14).stroke(Color(hex: "#f3c6da"), lineWidth: 1))
    }
}

/// One tile open: every version, newest first, tappable for a full-screen
/// look. The current version is labeled.
private struct AlbumTileView: View {
    let tile: APIClient.AlbumTile
    @State private var zoomed: APIClient.AlbumEntry?

    private let columns = [GridItem(.adaptive(minimum: 110), spacing: 8)]

    var body: some View {
        ScrollView {
            LazyVGrid(columns: columns, spacing: 8) {
                if let cur = tile.current { entry(cur, isCurrent: true) }
                ForEach(tile.history) { entry($0, isCurrent: false) }
            }
            .padding(12)
        }
        .background(Color(hex: "#fff7fb"))
        .navigationTitle(tile.label ?? "Tile")
        .navigationBarTitleDisplayMode(.inline)
        .sheet(item: $zoomed) { e in
            VStack(spacing: 14) {
                MediaImage(blobKey: e.blobKey)
                    .aspectRatio(1, contentMode: .fit)
                    .clipShape(RoundedRectangle(cornerRadius: 20))
                    .padding(.horizontal, 16)
                Text(tile.label ?? "")
                    .font(.system(size: 22, weight: .bold, design: .rounded))
                    .foregroundStyle(Color(hex: "#ad1457"))
                Text(dateText(e.when))
                    .font(.footnote).foregroundStyle(.secondary)
            }
            .padding(.vertical, 24)
            .presentationDetents([.large])
        }
    }

    private func entry(_ e: APIClient.AlbumEntry, isCurrent: Bool) -> some View {
        Button { zoomed = e } label: {
            VStack(spacing: 3) {
                ZStack(alignment: .topTrailing) {
                    MediaImage(blobKey: e.blobKey)
                        .frame(minWidth: 110, minHeight: 110)
                        .aspectRatio(1, contentMode: .fill)
                        .clipShape(RoundedRectangle(cornerRadius: 12))
                    if isCurrent {
                        Text("Current")
                            .font(.system(size: 9, weight: .bold))
                            .foregroundStyle(.white)
                            .padding(.horizontal, 6).padding(.vertical, 2)
                            .background(Color(hex: "#16a34a"), in: Capsule())
                            .padding(6)
                    }
                }
                Text(dateText(e.when))
                    .font(.system(size: 9))
                    .foregroundStyle(.secondary)
            }
        }
        .buttonStyle(.plain)
    }

    private func dateText(_ iso: String?) -> String {
        guard let iso else { return "" }
        let f = ISO8601DateFormatter()
        f.formatOptions = [.withInternetDateTime, .withFractionalSeconds]
        let d = f.date(from: iso) ?? ISO8601DateFormatter().date(from: iso)
        guard let d else { return "" }
        return d.formatted(date: .abbreviated, time: .omitted)
    }
}

File: android-native/app/src/main/java/io/andrewpeterson/myworld/ui/parent/AlbumView.kt
package io.andrewpeterson.myworld.ui.parent

import android.graphics.Bitmap
import androidx.compose.foundation.Image
import androidx.compose.foundation.background
import androidx.compose.foundation.border
import androidx.compose.foundation.clickable
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.aspectRatio
import androidx.compose.foundation.layout.fillMaxSize
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.height
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.width
import androidx.compose.foundation.lazy.grid.GridCells
import androidx.compose.foundation.lazy.grid.LazyVerticalGrid
import androidx.compose.foundation.lazy.grid.items
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.foundation.verticalScroll
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.runtime.LaunchedEffect
import androidx.compose.runtime.getValue
import androidx.compose.runtime.mutableStateOf
import androidx.compose.runtime.produceState
import androidx.compose.runtime.remember
import androidx.compose.runtime.setValue
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.draw.clip
import androidx.compose.ui.graphics.Color
import androidx.compose.ui.graphics.asImageBitmap
import androidx.compose.ui.layout.ContentScale
import androidx.compose.ui.text.font.FontWeight
import androidx.compose.ui.unit.Dp
import androidx.compose.ui.unit.dp
import androidx.compose.ui.unit.sp
import androidx.compose.ui.window.Dialog
import androidx.compose.ui.window.DialogProperties
import io.andrewpeterson.myworld.LocalAppContainer
import io.andrewpeterson.myworld.net.AlbumEntry
import io.andrewpeterson.myworld.net.AlbumTile
import io.andrewpeterson.myworld.net.albumByTile
import io.andrewpeterson.myworld.ui.theme.Brand
import io.andrewpeterson.myworld.ui.theme.hexColor
import java.time.Instant
import java.time.ZoneId
import java.time.format.DateTimeFormatter

/**
 * Memorabilia album — port of `Parent/AlbumView.swift`:
 * Album → folder (People / Words / Verbs / Celebrations) → tile → every
 * version newest-first, tappable for a big look. Old art is kept forever.
 */

private enum class AlbumFolder(val title: String, val emoji: String) {
    PEOPLE("People", "🧑‍🤝‍🧑"),
    WORDS("Words", "🏷"),
    VERBS("Verbs", "🏃"),
    CELEBRATIONS("Celebrations", "✨");

    fun matches(section: String?): Boolean {
        val s = (section ?: "").lowercase()
        return when (this) {
            PEOPLE -> s == "people"
            WORDS -> s == "nouns" || s == "needs" || s.isEmpty()   // unfiled lands here
            VERBS -> s == "verbs"
            CELEBRATIONS -> s == "events"
        }
    }
}

/** Small blob-key → bitmap image, shared by the parent screens. */
@Composable
fun BlobImage(blobKey: String?, modifier: Modifier = Modifier, contentScale: ContentScale = ContentScale.Crop) {
    val c = LocalAppContainer.current
    val bmp by produceState<Bitmap?>(initialValue = null, blobKey) {
        value = if (blobKey.isNullOrEmpty()) null
        else kotlinx.coroutines.withContext(kotlinx.coroutines.Dispatchers.Default) { c.media.bitmap(blobKey) }
    }
    val b = bmp
    if (b != null) {
        Image(b.asImageBitmap(), contentDescription = null, contentScale = contentScale, modifier = modifier)
    } else {
        Box(modifier.background(hexColor("#fce4ec")))
    }
}

@Composable
fun AlbumView(onDismiss: () -> Unit) {
    val c = LocalAppContainer.current
    var tiles by remember { mutableStateOf<List<AlbumTile>?>(null) }
    var error by remember { mutableStateOf<String?>(null) }
    var openFolder by remember { mutableStateOf<AlbumFolder?>(null) }
    var openTile by remember { mutableStateOf<AlbumTile?>(null) }
    var zoomed by remember { mutableStateOf<AlbumEntry?>(null) }

    LaunchedEffect(Unit) {
        try { tiles = c.api.albumByTile(c.auth.childSlug, limit = 600) }
        catch (e: Exception) { error = "Could not load the album: ${e.message}" }
    }

    Dialog(onDismissRequest = onDismiss, properties = DialogProperties(usePlatformDefaultWidth = false)) {
        Column(Modifier.fillMaxSize().background(hexColor("#fff7fb"))) {
            Row(
                Modifier.fillMaxWidth().background(Brand.pink).padding(horizontal = 14.dp, vertical = 12.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                Text(
                    when {
                        zoomed != null || openTile != null -> "‹ Back"
                        openFolder != null -> "‹ Back"
                        else -> "✕"
                    },
                    fontSize = 17.sp, fontWeight = FontWeight.Bold, color = Color.White,
                    modifier = Modifier.clickable {
                        when {
                            zoomed != null -> zoomed = null
                            openTile != null -> openTile = null
                            openFolder != null -> openFolder = null
                            else -> onDismiss()
                        }
                    }.padding(6.dp),
                )
                Spacer(Modifier.width(10.dp))
                Text(
                    openTile?.label ?: openFolder?.title ?: "Album",
                    fontSize = 18.sp, fontWeight = FontWeight.Bold, color = Color.White,
                )
            }

            val t = tiles
            when {
                error != null -> Text(error!!, fontSize = 13.sp, color = Color(0xFFDC2626),
                    modifier = Modifier.padding(16.dp))
                t == null -> LoadingSpinner("Loading…")
                t.isEmpty() -> Text(
                    "No pictures yet. As tiles get new art, the old versions are kept here forever.",
                    fontSize = 13.sp, color = Brand.muted, modifier = Modifier.padding(24.dp),
                )
                openTile != null -> TileVersionsGrid(openTile!!) { zoomed = it }
                openFolder != null -> FolderTileList(
                    t.filter { openFolder!!.matches(it.section) },
                ) { openTile = it }
                else -> FolderList(t) { openFolder = it }
            }
        }
    }

    zoomed?.let { e ->
        Dialog(onDismissRequest = { zoomed = null }) {
            Column(
                Modifier.fillMaxWidth().background(Color.White, RoundedCornerShape(24.dp)).padding(18.dp),
                horizontalAlignment = Alignment.CenterHorizontally,
            ) {
                BlobImage(e.blobKey, Modifier.fillMaxWidth().aspectRatio(1f).clip(RoundedCornerShape(20.dp)),
                    contentScale = ContentScale.Fit)
                Spacer(Modifier.height(10.dp))
                Text(openTile?.label ?: "", fontSize = 22.sp, fontWeight = FontWeight.Bold, color = Brand.pinkDeep)
                Text(albumDate(e.whenAt), fontSize = 13.sp, color = Brand.muted)
            }
        }
    }
}

@Composable
private fun FolderList(tiles: List<AlbumTile>, onOpen: (AlbumFolder) -> Unit) {
    Column(Modifier.fillMaxSize().verticalScroll(rememberScrollState()).padding(16.dp)) {
        AlbumFolder.entries.forEach { folder ->
            val inFolder = tiles.filter { folder.matches(it.section) }
            if (inFolder.isEmpty()) return@forEach
            val pictures = inFolder.sumOf { 1 + it.history.size }
            Row(
                Modifier.fillMaxWidth()
                    .background(Color.White, RoundedCornerShape(18.dp))
                    .border(1.dp, hexColor("#f3c6da"), RoundedCornerShape(18.dp))
                    .clickable { onOpen(folder) }
                    .padding(14.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                // Cover: the newest tile's art (the iOS fan collapses to one here).
                val cover = inFolder.firstOrNull()?.let { it.current?.blobKey ?: it.history.firstOrNull()?.blobKey }
                if (cover != null) {
                    BlobImage(cover, Modifier.size(60.dp).clip(RoundedCornerShape(10.dp)))
                } else {
                    Box(Modifier.size(60.dp).background(hexColor("#fce4ec"), RoundedCornerShape(10.dp)),
                        contentAlignment = Alignment.Center) { Text(folder.emoji, fontSize = 24.sp) }
                }
                Spacer(Modifier.width(14.dp))
                Column(Modifier.weight(1f)) {
                    Text(folder.title, fontSize = 18.sp, fontWeight = FontWeight.Bold, color = Brand.ink)
                    Text("${inFolder.size} tile${if (inFolder.size == 1) "" else "s"} · " +
                        "$pictures picture${if (pictures == 1) "" else "s"}",
                        fontSize = 12.sp, color = Brand.muted)
                }
                Text("›", fontSize = 20.sp, color = Brand.muted)
            }
            Spacer(Modifier.height(12.dp))
        }
    }
}

@Composable
private fun FolderTileList(tiles: List<AlbumTile>, onOpen: (AlbumTile) -> Unit) {
    Column(Modifier.fillMaxSize().verticalScroll(rememberScrollState()).padding(16.dp)) {
        tiles.forEach { tile ->
            Row(
                Modifier.fillMaxWidth()
                    .background(Color.White, RoundedCornerShape(14.dp))
                    .border(1.dp, hexColor("#f3c6da"), RoundedCornerShape(14.dp))
                    .clickable { onOpen(tile) }
                    .padding(12.dp),
                verticalAlignment = Alignment.CenterVertically,
            ) {
                val key = tile.current?.blobKey ?: tile.history.firstOrNull()?.blobKey
                BlobImage(key, Modifier.size(56.dp).clip(RoundedCornerShape(12.dp)))
                Spacer(Modifier.width(12.dp))
                Column(Modifier.weight(1f)) {
                    Text(tile.label ?: "Untitled", fontSize = 16.sp,
                        fontWeight = FontWeight.SemiBold, color = Brand.ink)
                    val total = 1 + tile.history.size
                    Text("$total picture${if (total == 1) "" else "s"}", fontSize = 12.sp, color = Brand.muted)
                }
                Text("›", fontSize = 18.sp, color = Brand.muted)
            }
            Spacer(Modifier.height(10.dp))
        }
    }
}

@Composable
private fun TileVersionsGrid(tile: AlbumTile, onZoom: (AlbumEntry) -> Unit) {
    val entries = buildList {
        tile.current?.let { add(it to true) }
        tile.history.forEach { add(it to false) }
    }
    LazyVerticalGrid(
        columns = GridCells.Adaptive(minSize = 110.dp),
        modifier = Modifier.fillMaxSize().padding(12.dp),
    ) {
        items(entries, key = { it.first.blobKey + (it.first.whenAt ?: "") }) { (e, isCurrent) ->
            Column(Modifier.padding(4.dp).clickable { onZoom(e) }) {
                Box {
                    BlobImage(e.blobKey, Modifier.fillMaxWidth().aspectRatio(1f).clip(RoundedCornerShape(12.dp)))
                    if (isCurrent) {
                        Text("Current", fontSize = 9.sp, fontWeight = FontWeight.Bold, color = Color.White,
                            modifier = Modifier.align(Alignment.TopEnd).padding(6.dp)
                                .background(hexColor("#16a34a"), RoundedCornerShape(50))
                                .padding(horizontal = 6.dp, vertical = 2.dp))
                    }
                }
                Text(albumDate(e.whenAt), fontSize = 9.sp, color = Brand.muted)
            }
        }
    }
}

private fun albumDate(iso: String?): String = try {
    if (iso == null) "" else Instant.parse(iso).atZone(ZoneId.systemDefault())
        .format(DateTimeFormatter.ofPattern("MMM d, yyyy"))
} catch (_: Exception) { "" }
